### ISSUE TITLE
Anonymous namespaces treated incorrectly with EXTRACT_ANON_NSPACES   = NO (Origin: bugzilla #679131)

### ISSUE THREAD

**Issue by doxygen** [doxygen/doxygen#4778]
status NEW severity *normal* in component *build* for *---*
Reported in version *1.8.1.1* on platform *Other*
Assigned to: Dimitri van Heesch

Original attachment names and IDs:
- _[namespaces.html](http://ftp.stack.nl/pub/users/dimitri/bugzilla/attachments//bug679131_217605_namespaces.html)_ (ID 217605)

On 2012-06-29 12:21:20 +0000, Simon Platt wrote:

> Created attachment 217605
> Faulty html output showing anonymous namespace treated incorrectly.
> 
> A file containing this code:
> 
> /** @file
> This is a source file for testing doxygen.
> */
> 
> namespace {
> 	int foo;	///< A silly name.
> 	double bar;	///< Another silly name.
> }
> 
> leads to the anonymous namespace being documented even with 
> EXTRACT_ANON_NSPACES   = NO
> as illustrated by namespaces.html, attached. 
> 
> Furthermore, the namespace appears in the namespace list as @n (@1 in namespaces.html; I have seen this several times with various numbers from @0 upwards).
> 
> Behaviour is as expected with
> EXTRACT_ANON_NSPACES   = YES

**Comment by kinxer** [comment]
This issue is still valid for 1.8.16.

**Comment by doxygen** [comment]
@kinxer Is it fixed with version 1.8.17? I couldn't reproduce this with the above example, so please attach a self-contained example (source+config file in a zip).

**Comment by kinxer** [comment]
I didn't see any references to it in the commit notes for 1.8.17, so I assumed it hadn't been fixed. I'll try 1.8.17.

**Comment by kinxer** [comment]
On Doxygen 1.8.17, compiled from source on OpenSUSE Tumbleweed 20200111, the example above outputs namespace XML for the example file given above and the following configuration:
```
GENERATE_XML = YES

INPUT = .
FILE_PATTERNS = *.cpp
EXTRACT_ANON_NSPACES   = NO
```
(where the test file is saved in the same directory as the doxyfile with a `.cpp` extension)

**Comment by kinxer** [comment]
Here is a zip with the example:
[doxygen-namespaces-example.zip](https://github.com/doxygen/doxygen/files/4060410/doxygen-namespaces-example.zip)

**Comment by kinxer** [comment]
I was able to patch this by making a change in `src/xmlgen.cpp`.

The current code, as of Release 1.8.17, is [here](https://github.com/doxygen/doxygen/blob/Release_1_8_17/src/xmlgen.cpp#L1997-L2003).

I changed that code to
```cpp
NamespaceSDict::Iterator nli(*Doxygen::namespaceSDict);
const NamespaceDef *nd;
for (nli.toFirst();(nd=nli.current());++nli)
{
  if (nd->isLinkable()) // added line
  { // added line
    msg("Generating XML output for namespace %s\n", nd->name().data());
    generateXMLForNamespace(nd, t);
  } // added line
}
```

This gives me the desired result (no output XML files for anonymous namespaces). However, in terms of code cleanliness, it seems like it would be preferable to prevent the anonymous namespace from being added to `Doxygen::namespaceSDict` at all.

**Comment by flynneva** [comment]
Any update on this issue? This is still an issue with v1.9.6.

**Comment by albert-github** [comment]
No there is no update yet, though I had a closer look (also with the current master version (1.9.8 (d02b813cd8241a74d64d426ba76b45c6849fdce4)).
The XML output should output all the information as available more or less irrespective of the doxygen settings and thus the anonymous namespace should be present, but also the members of the anonymous namespace. The later though is not the case and they should (I think be added).

@doxygen correct?

**Comment by albert-github** [comment]
I just looked in the file xmlgen.cpp in the function `generateXMLForClass` and here I see:
```
  if (cd->isReference())        return; // skip external references.
  if (cd->isHidden())           return; // skip hidden classes.
  if (cd->isAnonymous())        return; // skip anonymous compounds.
  if (cd->templateMaster()!=0)  return; // skip generated template instances.
  if (cd->isArtificial())       return; // skip artificially created classes
```
so probably the same reasoning is valued for the function generateXMLForNamespace where we now only see:
```
if (nd->isReference() || nd->isHidden()) return; // skip external references
```

**Comment by doxygen** [comment]
> The XML output should output all the information as available more or less irrespective of the doxygen settings and thus the anonymous namespace should be present, but also the members of the anonymous namespace. The later though is not the case and they should (I think be added).
>
> @doxygen correct?

Yes, the idea is that the XML output is a structured dump of the information collected by doxygen. So the options that control which output to show for other output formats do not apply for XML.

**Comment by albert-github** [comment]
@doxygen 
- isn't it a bit strange that the anonymous classes are excluded?
- so the variables should also be shown in the anonymous name space.
